Add Computus in Javascript and Typescript

hugo-s29 opened the pull request:
A translation of the computus Julia code in javascript and in typescript.

ntindle commented on contents/computus/code/typescript/gauss_easter.ts:
```diff
@@ -0,0 +1,80 @@
+
+function computus(year: number, servois: boolean = false): string {
+  // Year's position in metonic cycle
+  const a = year % 19;
+
+  // Century index
+  const k = Math.floor(year / 100);
+
+  // Shift of metonic cycle, add a day offset every 300 years
+  const p = Math.floor((13 + 8 * k) / 25);
+
+  // Correction for non-observed leap days
+  const q = Math.floor(k / 4);
+
+  // Correction to starting point of calculation each century
+  const M = mod(15 - p + k - q, 30);
+
+  // Number of days from March 21st until the full moon
+  const d = (19 * a + M) % 30;
```
Any reason you aren't explicitly typing these?

hugo-s29 replied:
I didn't explicitly type these variables to be clearer to read. 
However, it may be better with types: if someone wants an implementation without types, they could use the JavaScript version. 
I'll add types to every variable as soon as possible.

ntindle replied:
great, thanks!

ntindle commented on contents/computus/code/typescript/gauss_easter.ts:
```diff
@@ -0,0 +1,80 @@
+  // Days from d to next Sunday
+  let e = (2 * b + 4 * c + 6 * d + N) % 7;
+
+  // Historical corrections for April 26 and 25
+  if (e === 6) if (d === 29 || (d === 28 && a > 10)) e = -1;
```
Would you mind bringing the reassignment into its own line? It's a bit buried at the end as is.

hugo-s29 replied:
Yes, I can fix that.

ntindle replied:
thanks, would you mind doing that for all instances of inline if/else

hugo-s29 commented on contents/computus/computus.md:
```diff
@@ -320,6 +320,10 @@ For now, we have the code outputting a tuple of $$d$$ and $$e$$, so users can us
 {% sample lang="nim" %}
 [import, lang:"nim"](code/nim/gauss_easter.nim)
+{% sample lang="javascript" %}
+[import, lang:"nim"](code/javascript/gauss_easter.js)
```
I'll also need to change the lang "variable" to JavaScript and TypeScript. I forgot to do that.

ntindle replied:
great catch, I missed this as well

hugo-s29 commented:
I added types to the typescript code.
I changed the formatting for javascript and typescript.
I added a Dart implementation.

Amaras commented:
I think you're trying to go a little too fast with the PRs. 😅
I'd recommend separating your Dart implementation from the TypeScript and JavaScript ones, because I don't know if we have competent Dart reviewers, so that might prevent your current PR from getting merged.
It's not really that important for the Computus because it's so simple, but keep that in mind in the future.

leios commented:
Since you are a first time contributor, feel free to add your name to the list in CONTRIBUTORS.md.

Also: Dart seems to be a new language in the AAA, could you add a section to the book.json and make sure syntax highlighting is correct? It should be, but it would be nice to double-check!

leios commented:
For this PR, the implementation is so straightforward, that I feel it is ok to merge Dart as long as it produces the right results; however, definitely split it in future PRs.

Thanks again for the submission!

ntindle commented:
JS Output:
![JS Output](https://user-images.githubusercontent.com/8845353/130553678-fff3030d-1b8e-4791-ac23-1d15e24ec720.png)
TS Output:
![TS Output](https://user-images.githubusercontent.com/8845353/130553701-77e16d5a-415c-4dd6-b20d-59492f6822a1.png)

ntindle commented:
Dart output:
![Dart Output](https://user-images.githubusercontent.com/8845353/130553895-1bb50c8c-e597-4ad8-9e9e-5c7461acee67.png)

ntindle commented:
Note still needs lang highlighting for dart like #839 book.json

Amaras reviewed (approved):
If @ntindle is okay with it, I think it's okay with me.

hugo-s29 commented:
I added my name to the `CONTRIBUTORS.md` and added the Dart language to the `book.json`.
The syntax highlighting for Dart is working, I checked it.

Also, `contents/convolutions/res/Sobel_filters.png` and `contents/convolutions/res/sobel_filters.png` were conflicting when working on Windows, so I had to delete one.

About the Dart implementation, I could have created another PR. Sorry about that.

Amaras commented:
Right, I think we didn't consider Windows' path system. From manual inspection, it looks like they are the same images, though, so I don't think it's that much of a problem. We'll have to see what breaks, and make sure there are no case-dependant paths in the AAA any more for future-proofing it.
Apart from the conflicting images, I don't have anything against your PR, I think it's ready to merge.
I think I'll let someone else merge this PR, because we'll probably try to re-include the missing image before merging, and I'm not entirely competent. :sweat_smile: 

Maybe you could open an issue and/or a PR to correct the name conflict problem, if you feel like it. It would be appreciated. :wink:

hugo-s29 commented:
It seems the only file referencing this image is `contents/convolutions/2d/2d.md` and it only uses `sobel_filters.png`. I'll open a pull request for that. 
Otherwise, I've readded the missing file.
